scripts/lib/mkcloud-dirmaint.sh

```diff
@@ -19,57 +21,77 @@ function dirmaint_do_sanity_checks()
     fi
 }
 
+function dirmaint_user_deleted()
+{
+    vmcp q $1 2> /dev/null | grep HCPCQV003E >& /dev/null
+    return $?
+}
+
 function dirmaint_do_shutdowncloud()
 {
     for i in $(nodes ids all); do
-        vmcp sig shut $(printf "${cloud}n%02d" $i)
+        # skip user, if it is logged off or if it does not exist
+        vmcp q $(printf "${cloud}n%02d" $i) 2> /dev/null |\
+            grep -e "HCPCQU045E" -e "HCPCQV003E" >& /dev/null && continue
+        sigcmd="cp sig shut $(printf "${cloud}n%02d" $i) within 60"
+        forcecmd="cp force $(printf "${cloud}n%02d" $i)"
+        $lxc -c "$sigcmd" || lxc -c "$forcecmd"
     done
-    vmcp sig shut ${cloud}adm
-    wait_for 60 1 "! vmcp q ${cloud}adm" "admin node ${cloud}adm to log off"
+    # if a guest is logged on, query to the name has status 0
+    vmcp q ${cloud}adm >& /dev/null && $lxc -c "cp force ${cloud}adm IMMED"
+    vmcp q ${cloud}adm >& /dev/null && wait_for 60 1 "! vmcp q ${cloud}adm" "admin node ${cloud}adm to log off"
+    # finally remove cloud administration guest from directory:
+    $lxc -c "dirm for ${cloud}adm purge"
+    # wait until the directory entry is deleted
+    # to improve performance, we might want to remove the disk from the admin node first and purge then.
+    wait_for 120 1 "dirmaint_user_deleted ${cloud}adm" "admin node ${cloud}adm to be purged (disk deletion)"
 }
 
 function dirmaint_do_cleanup()
 {
+    local vdev="0a${cloudidx}0"
+    echo "cleaning up for locally linked minidisk $vdev"
+    # cleanup if copy of image to admin node or dasdfmt aborted failed
+    cat /proc/dasd/devices | grep "0.0.$vdev"  && chccwdev -d 0.0.$vdev
+    vmcp q v $vdev >& /dev/null && vmcp detach $vdev
+    # now we can proceed to try and cleanup the rest
     dirmaint_do_shutdowncloud
 
     killproc -p /var/run/mkcloud/dnsmasq-$cloud.pid /usr/sbin/dnsmasq
     rm -f /var/run/mkcloud/dnsmasq-$cloud.pid /etc/dnsmasq-$cloud.conf
 
     ndev=1${cloudidx}00
-    devname=$(cat /sys/bus/ccwgroup/drivers/qeth/0.0.${ndev}/if_name)
-
-    # take network down for this cloud
-    ip link set $devname down
-    echo 0 > /sys/bus/ccwgroup/drivers/qeth/0.0.${ndev}/online
-    vmcp uncouple ${ndev}
-    vmcp detach nic ${ndev}
-    vmcp set nic ${ndev} macid system
+    if [ -f /sys/bus/ccwgroup/drivers/qeth/0.0.${ndev}/if_name ]; then
+        devname=$(cat /sys/bus/ccwgroup/drivers/qeth/0.0.${ndev}/if_name)
 
+        # take network down for this cloud
+        ip link set $devname down
+        # use system script to take down the device
+        qeth_configure -l 0.0.1${cloudidx}00 0.0.1${cloudidx}01 0.0.1${cloudidx}02 0
+        $lxc -c "cp for lnxadmin cmd detach nic ${ndev}"
```

Comment:
lnxadmin is the machine where this script is running. I probably should check for the z/VM name of the current machine and just use it...